Log collector

paweldomas opened the pull request:
Adds LogCollector which can be used to capture and store logs when added as global transport.

paweldomas committed:
fix(Logger): logging level comparision

Before the fix the logger would not log when the log level
equals it's level.

paweldomas committed:
feat: add global transport

Global transport can be registered via static method
and will receive messages from all logger instances.

paweldomas committed:
feat: add LogCollector

paweldomas committed:
feat(LogCollector): aggregate the same messages

hristoterezov commented:
@paweldomas Currently the project is written with ES5 syntax. 

I noticed that you are using `const` from time to time. If you want to adopt the ES6 syntax in the project (which is great idea!) why are you using only `const`. Maybe your new class can be defined with `class` and not with the `prototype`, you can use `let`, arrow functions, etc. Also if you want to do this maybe it's good idea to include babel in the package JSON as devDependacy.

If you don't want to do this now I think it's going to be better to remove the usage of `const`.

paweldomas committed:
fix: remove 'const'

paweldomas commented:
@hristoterezov thanks for suggestions ! I removed the "const" for now.

hristoterezov commented on lib/LogCollector.js:
```diff
@@ -0,0 +1,231 @@
+ */
+function stringify(someObject){
+    const cache = [];
+    return JSON.stringify(someObject, function (key, value) {
```
I see the following problem with this implementation:
```
var obj = {a: 2}
stringify({a: obj, b: obj}) // Result: "{"a":{"a":2}}"
JSON.stringify({"a": obj, b: obj}) //Result: "{"a":{"a":2},"b":{"a":2}}"
```

My proposal is to use the original JSON.stringify and if somebody is trying to log something with circular dependancies to handle this on it's own. That way the person who is logging the message will have to decide how to resolve the circular dependency.

paweldomas replied:
We could leave it for the library users, but since we need to solve this problem anyway then maybe I'll try to fix it.

paweldomas replied:
The problem is quite complex and should probably be solved when the "stringificator" has more knowledge about the object's structure.
I've added a special string which prints object's index from the cache list which allows to potentially identify objects in the structure. Given that the purpose of this stringification is mostly debugging I'd rather not go further than that: 
https://github.com/jitsi/jitsi-meet-logger/pull/1/commits/3e8e2bd147329784b0f1ee1b6d2898dbe444bd08

paweldomas replied:
Ok, added another one: https://github.com/jitsi/jitsi-meet-logger/pull/1/commits/39570a32b191b3a02d738aaeaf2a090abb13a124 will remove the previous one with force push

hristoterezov commented on lib/LogCollector.js:
```diff
@@ -0,0 +1,207 @@
+ *
+ * @constructor
+ */
+function LogCollector(logStorage, options) {
```
Isn't this class too specific to be included in jitsi-meet-logger? Isn't it better to be defined outside of jitsi-meet-logger?

paweldomas replied:
It feels like it can be useful utility for other people using logger, but I don't mind moving it somewhere else. Should I move it to jitsi-meet ? But then it doesn't fit there very well, because will be used directly from jitsi-meet like the logger is (jitsi-meet-logger is dependency of both jitsi-meet and lib-jitsi-meet).

hristoterezov commented on lib/LogCollector.js:
```diff
@@ -0,0 +1,207 @@
+ * Starts this instance and schedules periodical "store logs" task.
+ */
+LogCollector.prototype.start = function () {
+    this._reschedulePublishInterval();
```
Isn't it better if the start method is called automatically when the first log is available? 

If we make this change, probably the stop method should be renamed somehow (maybe dispose).

hristoterezov replied:
@paweldomas Please discard this comment. I saw the code in jitsi-meet and you actually need this to cache the first logs, right?

paweldomas replied:
it will cache the logs as soon as is added as global transport, but it not try to store them until start is called

paweldomas committed:
ref(LogCollector): move JSON formatting logic to the storage

Also separates methods for formatting and stringifying the log message
which could be possibly overridden by extending the class.

paweldomas committed:
doc(LogCollector): updates start and stop descriptions

paweldomas commented:
Hey Hristo, so as we discussed it yesterday "of the list" I've made some efforts to make LogCollector less specific. The last commit removes the logic that crafts JSON message and pushes it to the log storage implementation. It also extracts formatting and stringifying parts into methods, so that they could be overridden.

On the other hand I tried the other approach that we discussed and I extracted StringLogCollector from AbstractLogCollector which:
- assumes that it is the string that is being logged
- implements the condition whether the logs should be stored based on string length
- implements "same as the previous one" log message aggregation
A "sketch" of that solution can be found on the branch here:
https://github.com/jitsi/jitsi-meet-logger/tree/log_collector_abstract

I don't like this "abstract" approach, because:
1. logs are usually string, so I don't really see reason why would you want to do it different way
2. both format and stringifying can be overridden with the 1st approach found in this PR
3. AbstractLogCollector feels empty, the whole thing only adds complexity

paweldomas commented:
Note that I am still working on fixing the stringify method

paweldomas committed:
fix(LogCollector): use JSON.stringify with try catch

paweldomas committed:
feat(LogCollector): add caching before storage is ready
